### ISSUE TITLE
Error when EPServiceProviderManager.GetDefaultProvider();

### ISSUE THREAD

**Issue by juanmaotegui** [espertechinc/nesper#22]
I'm migrating esper java to nesper net core, in java works perfect. But now in net core, i'm having some troubles. In the line `EPServiceProvider serviceProvider = EPServiceProviderManager.GetDefaultProvider();` i'm getting [this error](https://imgur.com/a/5K7jaDx) 
i don't know why is throwing me this error, because i don't pass any parameter Month o something like that.
My code

        public class Events
        {
        public static Events Instance;
        private string Expression;
        private EPRuntime Runtime { get; set; }
        private EPAdministrator Administrator { get; set; }
        private EPStatement Statement { get; set; }
        private RulesHandler Handler { get; set; } = RulesHandler.GetInstance;

        private IDataAccessSprayerTruck _sprayerTruckDataAccess;
        private IDataAccessUser _userDataAccess;

        protected Events()
        {
            IServiceCollection services = new ServiceCollection();

            services.AddScoped<IDataAccessSprayerTruck, SprayerTruckDataAccess>();
            services.AddScoped<IDataAccessUser, UserDataAccess>();

            IServiceProvider provider = services.BuildServiceProvider();

            _sprayerTruckDataAccess = provider.GetService<IDataAccessSprayerTruck>();
            _userDataAccess = provider.GetService<IDataAccessUser>();

            InitializeEsper();
        }

        public static Events GetInstance
        {
            get
            {
                if (Instance == null)
                    Instance = new Events();

                return Instance;
            }
        }

        private void InitializeEsper()
        {
            try
            { 
                EPServiceProvider serviceProvider = EPServiceProviderManager.GetDefaultProvider(EventConfig());
                Runtime = serviceProvider.EPRuntime;
                Administrator = serviceProvider.EPAdministrator;
                Administrator.Configuration.AddEventType<Record>();
            }
            catch (Exception ex)
            {
                Console.WriteLine("ERROR --> InitializeEsper: " + ex.Message);
            }
        }

        private static Configuration EventConfig()
        {
            Configuration config = new Configuration();
            config.AddEventType<Record>();
            return config;
        }

        public void GenerateEvent(Record record)
        {
            if (record != null)
            {
                if (Expression != null)
                {
                    Runtime.SendEvent(record);
                }
            }
        }

        }

I call the method `GenerateEvent`

        private Events Events { get; set; } = Events.GetInstance;
        Events.GenerateEvent(record);

**Comment by fcabrera23** [comment]
@ajaxx this is the problem we are getting while trying to set up nesper. We are using net core 2.1

**Comment by ajaxx** [comment]
Interesting, if I had to guess it probably means there is a TimeZoneInfo somewhere in the serialization data.  I don't see anything specific about your code that looks like it should cause a serialization issue though.  I'll take this code and see if I can build it into something on my end and reproduce the issue.

**Comment by fcabrera23** [comment]
We tried the same code on windows machine and worked. The problem is with when using net core in Mac. Please feel free to suggest or ask me to try what you believe could be the error

**Comment by fcabrera23** [comment]
Hi @ajaxx  I have just tried the same code using Ubuntu 16.04 and its woking ok, so I can imagine this is a macOS problem.

**Comment by fcabrera23** [comment]
Hi @ajaxx  we are still having problem on Ubuntu and macOS. 

Now we are using a basic example, provided from espertech website
using System;
using System.Threading;
using com.espertech.esper.client;
namespace PruebaNesper
{
    class Program
    {
        private static Timer _eventTimer;
        private static Random _random;

        private static EPAdministrator _administrator;
        private static EPRuntime _runtime;



        static void Main(string[] args)
        {
            InitializeEsper();
            InitializeEvents();

            // Make the application hang around - failure to do this will result
            // in the application terminating because there are no non-daemon
            // threads running.
            Console.WriteLine("Application waiting until you hit return to end");
            Console.ReadLine();

        }


        /// <summary>
        /// Initializes Esper and prepares event listeners.
        /// </summary>
        static void InitializeEsper()
        {
            var serviceProvider = EPServiceProviderManager.GetDefaultProvider();
            _runtime = serviceProvider.EPRuntime;
            _administrator = serviceProvider.EPAdministrator;
            // You must tell esper about your events, failure to do so means you will be using
            // the fully qualified names of your events.  There are many ways to do that, but
            // this one is short-hand which makes the name of your event aliased to the fully
            // qualified name of your type.
            _administrator.Configuration.AddEventType<TradeEvent>();
            // Create a statement or pattern; these are the bread and butter of esper.  This
            // method creates a statement.  You want to hang on to the statement if you intend
            // to listen directly to the results.
            var statement = _administrator.CreateEPL("select * from TradeEvent");
            // Hook up an event handler to the statement
            statement.Events += new UpdateEventHandler(HandleEvent);
        }

        /// <summary>
        /// Handles the event.
        /// </summary>
        /// <param name="sender">The sender.</param>
        /// <param name="args">The <see cref="com.espertech.esper.client.UpdateEventArgs"/> instance containing the event data.</param>
        private static void HandleEvent(object sender, UpdateEventArgs args)
        {
            // if you have a window or a complex rule, you will find that the event args contains
            // information about events that are no longer applicable and those that are currently
            // applicable.  For now, we only care about those events that are directly applicable.
            foreach (var e in args.NewEvents)
            {
                // e is an "EventBean" which is a wrapper around the underlying object.  If you
                // are dealing with a native object, you can access it by looking directly at
                // the underlying as follows.
                Console.WriteLine("TradeEvent - " + e.Underlying);
                // However, more complex types and maps don't necessarily behave as cleanly and
                // the EventBean hides some of that from you.
                Console.WriteLine("TradeEvent - Symbol = " + e.Get("Symbol"));
                Console.WriteLine("TradeEvent - Price = " + e.Get("Price"));
                Console.WriteLine("TradeEvent - Quantity = " + e.Get("Quantity"));
            }
        }

        /// <summary>
        /// Esper requires events to come from somewhere; this application is just a dummy
        /// application so we have to manufacture the events.  In a real application, you
        /// would receive these from "something" that produces these events.
        /// </summary>
        static void InitializeEvents()
        {
            _random = new Random();
            _eventTimer = new Timer(
                o => SendTradeEvent(),
                null,
                100,
                100);

        }

        /// <summary>
        /// Sends the trade event.
        /// </summary>
        static void SendTradeEvent()
        {
            // Create a phantom event
            var tradeEvent = new TradeEvent()
            {
                Symbol = "GOOG",
                Price = 700 + Math.Round(_random.NextDouble() * 10, 1),
                Quantity = 100 + _random.Next(1, 10) * 100
            };

            // Events must be injected into the runtime
            _runtime.SendEvent(tradeEvent);
        }
    }

    public class TradeEvent
    {
        public string Symbol { get; set; }
        public int Quantity { get; set; }
        public double Price { get; set; }

        /// <summary>
        /// Returns a <see cref="T:System.String"/> that represents the current <see cref="T:System.Object"/>.
        /// </summary>
        /// <returns>
        /// A <see cref="T:System.String"/> that represents the current <see cref="T:System.Object"/>.
        /// </returns>
        /// <filterpriority>2</filterpriority>
        public override string ToString()
        {
            return string.Format("Symbol: {0}, Quantity: {1}, Price: {2}", Symbol, Quantity, Price);
        }
    }
}'


When I run the program using commando "dotnet run" I get the following error 

Unhandled Exception: System.Runtime.Serialization.SerializationException: An error occurred while deserializing the object.  The serialized data is corrupt. ---> System.ArgumentOutOfRangeException: The Month parameter must be in the range 1 through 12.
Parameter name: month
   at System.TimeZoneInfo.TransitionTime.ValidateTransitionTime(DateTime timeOfDay, Int32 month, Int32 week, Int32 day, DayOfWeek dayOfWeek)
   at System.TimeZoneInfo.TransitionTime.System.Runtime.Serialization.IDeserializationCallback.OnDeserialization(Object sender)
   --- End of inner exception stack trace ---
   at System.TimeZoneInfo.TransitionTime.System.Runtime.Serialization.IDeserializationCallback.OnDeserialization(Object sender)
   at System.Runtime.Serialization.ObjectManager.RaiseDeserializationEvent()
   at System.Runtime.Serialization.Formatters.Binary.ObjectReader.Deserialize(BinaryParser serParser, Boolean fCheck)
   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize(Stream serializationStream, Boolean check)
   at com.espertech.esper.util.SerializableObjectCopier.Copy(IContainer container, Object orig)
   at com.espertech.esper.core.service.EPServiceProviderImpl.TakeSnapshot(Configuration configuration)
   at com.espertech.esper.core.service.EPServiceProviderImpl..ctor(IContainer container, Configuration configuration, String engineURI, IDictionary`2 runtimes)
   at com.espertech.esper.client.EPServiceProviderManager.GetProvider(IContainer container, String providerURI, Configuration configuration)
   at PruebaNesper.Program.InitializeEsper() in /home/ubuntu/PruebaNesper/PruebaNesper/Program.cs:line 36
   at PruebaNesper.Program.Main(String[] args) in /home/ubuntu/PruebaNesper/PruebaNesper/Program.cs:line 19

**Comment by bernhardttom** [comment]
Did you try to run regression tests at https://github.com/espertechinc/nesper/tree/master/NEsper/NEsper.Regression in the same configuration as the tests.

**Comment by fcabrera23** [comment]
NEsper project gets the following error -> /Users/franciscocabrera/Downloads/nesper-master/NEsper/NEsper/NEsper.csproj: Error NU1202: Package System.Configuration.ConfigurationManager 4.4.1 is not compatible with net452 (.NETFramework,Version=v4.5.2). Package System.Configuration.ConfigurationManager 4.4.1 supports:
  - net461 (.NETFramework,Version=v4.6.1)
  - netstandard2.0 (.NETStandard,Version=v2.0) (NU1202) (NEsper)

**Comment by fcabrera23** [comment]
Hi @ajaxx @bernhardttom @DixonDs  any news on this issue? Thanks

**Comment by pedrolopix** [comment]
I've got this problem too, in windows works fine and MacOS not. Any news?

**Comment by fcabrera23** [comment]
@pedrolopix we had no news regarding this issue. Right now we are using Java version until we fix the problem. @ajaxx do you have an estimated date? can we help with testing or debugging?

**Comment by ajaxx** [comment]
Good news is that EPServiceProvider is all but gone in 8.0.  The next good news is that I think we have an alpha version to begin test.  But there are a lot of changes and I'd like to make a request to have some alpha testers for 8.0.

**Comment by ajaxx** [comment]
FWIW, I'm completely open to providing a fix for this in 6.1.0 - I just needed to get through this 8.0 upgrade which has been a pretty substantial drain on me personally.  Any help with testing 8.0 is welcome (and I'd be incredibly grateful).

**Comment by fcabrera23** [comment]
@ajaxx  we can help with the 8.0 version. Its available to download via NuGet?

**Comment by fcabrera23** [comment]
Also, perhaps you can guide me on how to fix this for 6.1.0 and we can create a branch, fix it and then merge to current master branch.. do you agree?

**Comment by ajaxx** [comment]
I'm fine with creating a branch that contains a fix for this problem.  Sorry that I haven't had a chance to actively work on it like I'd prefer.  As for 8.0, I've created an 8.0-wip (work in progress) branch where most of my changes are... and it's pretty exhaustive so be forewarned.

**Comment by DarrenHart1989** [comment]
@ajaxx - Was there ever a branch created with a fix for this? Or is it just a case of moving to the WIP branch?

**Comment by ajaxx** [comment]
If you've looked at the wip branch, you've probably noticed some serious
structural changes. Unfortunately all of my time has been spent on the new
branch and I've not had any time to look at the current release branch. If
you have a patch send me a pr and I'll review and add it.

**Comment by o-mdr** [comment]
I suspect the issue is in the .net core, rather than in NEsper itself. @ajaxx if you still use `SerializableObjectCopier` anywhere in the `8.*` branch - i don't think it will work in .net core 2.2 (or netstandard2.0) on Linux (note, this works fine on Windows). 

For anyone who is interested, I've debugged this issue on an Ubuntu machine and narrowed it down to a short reproducible code: https://stackoverflow.com/q/57655469/706456.

```
using System;
using System.IO;
using System.Runtime.Serialization.Formatters.Binary;

namespace Test
{
    public class Program
    {
        public static void Main(string[] args)
        {
            // any zone here, don't care what it is
            var zone = TimeZoneInfo.GetSystemTimeZones()[0];            
            var formatter = new BinaryFormatter();

            using (MemoryStream stream = new MemoryStream())
            {
                formatter.Serialize(stream, zone);
                stream.Seek(0, SeekOrigin.Begin);
                var result = formatter.Deserialize(stream);
                Console.WriteLine("all ok");
            }
        }
    }
}
```

**Comment by ajaxx** [comment]
@o-mdr - any thoughts on alternatives to Serialization.  As a side note, I've abstracted the concept into a general purpose ObjectCopier which can be replace with any solution.  This will for obvious reasons break some compatibility with Java, but you should be able to register an object copier of your choice with the container framework.  I'll be testing this further once all (or most) NEsper 8 regression tests are operational.

**Comment by o-mdr** [comment]
Hi @ajaxx, there were two ways to address this particular issue:
- Use the newest .NET Core ([it was fixed](https://github.com/dotnet/corefx/issues/40578), but I don't know if/when it was released yet)
- There was a workaround for the serialiser, see this [stack overflow suggestion](https://stackoverflow.com/a/57662000/706456). All serializable `TimeZoneInfo`  properties will have to be changed to `get/set` notation with a backing field and `get` has to use `TimeZoneInfo.FindSystemTimeZoneById(TimeZoneId)` method. I have tested this workaround and it did work. The problem was that  `TimeZoneInfo` could not be deserialised on Linux platforms
